src/python/tritonfrontend/_c/tritonfrontend.h

```diff
@@ -0,0 +1,134 @@
+struct TritonError : public std::runtime_error {
+  explicit TritonError(const std::string& what) : std::runtime_error(what) {}
+};
+
+// triton::core::python exceptions map 1:1 to TRITONSERVER_Error_Code.
+struct UnknownError : public TritonError {
+  explicit UnknownError(const std::string& what) : TritonError(what) {}
+};
+struct InternalError : public TritonError {
+  explicit InternalError(const std::string& what) : TritonError(what) {}
+};
+struct NotFoundError : public TritonError {
+  explicit NotFoundError(const std::string& what) : TritonError(what) {}
+};
+struct InvalidArgumentError : public TritonError {
+  explicit InvalidArgumentError(const std::string& what) : TritonError(what) {}
+};
+struct UnavailableError : public TritonError {
+  explicit UnavailableError(const std::string& what) : TritonError(what) {}
+};
+struct UnsupportedError : public TritonError {
+  explicit UnsupportedError(const std::string& what) : TritonError(what) {}
+};
+struct AlreadyExistsError : public TritonError {
+  explicit AlreadyExistsError(const std::string& what) : TritonError(what) {}
+};
+
+void
+ThrowIfError(TRITONSERVER_Error* err)
+{
+  if (err == nullptr) {
+    return;
+  }
+  std::shared_ptr<TRITONSERVER_Error> managed_err(
+      err, TRITONSERVER_ErrorDelete);
+  std::string msg = TRITONSERVER_ErrorMessage(err);
+  switch (TRITONSERVER_ErrorCode(err)) {
+    case TRITONSERVER_ERROR_INTERNAL:
+      throw InternalError(std::move(msg));
+    case TRITONSERVER_ERROR_NOT_FOUND:
+      throw NotFoundError(std::move(msg));
+    case TRITONSERVER_ERROR_INVALID_ARG:
+      throw InvalidArgumentError(std::move(msg));
+    case TRITONSERVER_ERROR_UNAVAILABLE:
+      throw UnavailableError(std::move(msg));
+    case TRITONSERVER_ERROR_UNSUPPORTED:
+      throw UnsupportedError(std::move(msg));
+    case TRITONSERVER_ERROR_ALREADY_EXISTS:
+      throw AlreadyExistsError(std::move(msg));
+    default:
+      throw UnknownError(std::move(msg));
+  }
+}
+
+
+template <typename Base, typename FrontendServer>
+class TritonFrontend {
+ private:
+  std::shared_ptr<TRITONSERVER_Server> server_;
+  std::unique_ptr<Base> service;
+  triton::server::RestrictedFeatures restricted_features;
+
+ public:
+  TritonFrontend(uintptr_t server_mem_addr, UnorderedMapType data)
+  {
+    TRITONSERVER_Server* server_ptr =
+        reinterpret_cast<TRITONSERVER_Server*>(server_mem_addr);
+    server_.reset(server_ptr, DummyDeleter);
+
+    // For debugging
+    // for (const auto& [key, value] : data) {
+    //     std::cout << "Key: " << key << std::endl;
+    //     printVariant(value);
+    // }
+
+    ThrowIfError(
+        FrontendServer::Create(server_, data, restricted_features, &service));
+  };
+
+  void StartService() { ThrowIfError(service->Start()); };
+  void StopService() { ThrowIfError(service->Stop()); };
+
+
+  static TRITONSERVER_Error* DummyDeleter(TRITONSERVER_Server* obj)
+  {
+    return nullptr;  // Prevents double-free
```

Comment:
I'd be a little more descriptive here. Explain that the frontend doesn't own it and that we want to let the core binding handle the lifecycle (if that's correct)

Reply:
'empty' , null, Noop? instead of 'dummy'